File: src/services/conflict.ts
import { invoke } from "@tauri-apps/api/core";
import type { CommitResult } from "./git";

export interface ConflictBlock {
  ours: string;
  base: string | null;
  theirs: string;
  start_line: number;
  end_line: number;
}

export interface MergeBaseContent {
  path: string;
  base_content: string | null;
  ours_content: string;
  theirs_content: string;
}

export interface ConflictFile {
  path: string;
  conflict_count: number;
  conflicts: ConflictBlock[];
}

export type ConflictResolution =
  | { type: "Ours" }
  | { type: "Theirs" }
  | { type: "Both" }
  | { type: "Manual"; content: string };

export function getConflictFiles(tabId: string): Promise<ConflictFile[]> {
  return invoke<ConflictFile[]>("get_conflict_files", { tabId });
}

export function resolveConflict(
  tabId: string,
  path: string,
  resolution: ConflictResolution,
): Promise<void> {
  return invoke<void>("resolve_conflict", { tabId, path, resolution });
}

export function resolveConflictBlock(
  tabId: string,
  path: string,
  blockIndex: number,
  resolution: ConflictResolution,
): Promise<void> {
  return invoke<void>("resolve_conflict_block", {
    tabId,
    path,
    blockIndex,
    resolution,
  });
}

export function markResolved(tabId: string, path: string): Promise<void> {
  return invoke<void>("mark_resolved", { tabId, path });
}

export function abortMerge(tabId: string): Promise<void> {
  return invoke<void>("abort_merge", { tabId });
}

export function continueMerge(
  tabId: string,
  message: string,
): Promise<CommitResult> {
  return invoke<CommitResult>("continue_merge", { tabId, message });
}

export function isMerging(tabId: string): Promise<boolean> {
  return invoke<boolean>("is_merging", { tabId });
}

export function getMergeBaseContent(
  tabId: string,
  path: string,
): Promise<MergeBaseContent> {
  return invoke<MergeBaseContent>("get_merge_base_content", { tabId, path });
}

File: src-tauri/src/commands/conflict.rs
use tauri::State;

use crate::commands::with_repo;
use crate::git::types::{CommitResult, ConflictFile, ConflictResolution};
use crate::state::AppState;

#[tauri::command]
pub fn get_conflict_files(
    tab_id: String,
    state: State<'_, AppState>,
) -> Result<Vec<ConflictFile>, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.get_conflict_files().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn resolve_conflict(
    tab_id: String,
    path: String,
    resolution: ConflictResolution,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend
            .resolve_conflict(&path, resolution)
            .map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn resolve_conflict_block(
    tab_id: String,
    path: String,
    block_index: usize,
    resolution: ConflictResolution,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend
            .resolve_conflict_block(&path, block_index, resolution)
            .map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn mark_resolved(
    tab_id: String,
    path: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.mark_resolved(&path).map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn abort_merge(tab_id: String, state: State<'_, AppState>) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.abort_merge().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn continue_merge(
    tab_id: String,
    message: String,
    state: State<'_, AppState>,
) -> Result<CommitResult, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.continue_merge(&message).map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn is_merging(tab_id: String, state: State<'_, AppState>) -> Result<bool, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.is_merging().map_err(|e| e.to_string())
    })
}
